diff --git a/tools/swupdate-progress.c b/tools/swupdate-progress.c
index bd5a7913..0dd1779f 100644
--- a/tools/swupdate-progress.c
+++ b/tools/swupdate-progress.c
@@ -24,7 +24,7 @@
 #include <netinet/in.h>
 #include <pthread.h>
 #include <getopt.h>
-
+#include <curl/curl.h>
 #include <progress_ipc.h>
 
 #define PSPLASH_MSG_SIZE	64
@@ -238,6 +238,7 @@ int main(int argc, char **argv)
 	int opt_w = 0;
 	int opt_r = 0;
 	int opt_p = 0;
+	int opt_u = 0;
 	int c;
 	char *script = NULL;
 	char *reboot_script = NULL;
@@ -245,8 +246,14 @@ int main(int argc, char **argv)
 	bool disable_reboot = false;
 	bool redirected = false;
 
+	CURL *curl;
+	CURLcode res;
+	char put_buffer[1024];
+	struct curl_slist *put_headers = NULL;
+
+
 	/* Process options with getopt */
-	while ((c = getopt_long(argc, argv, "cwprhs:e:q",
+	while ((c = getopt_long(argc, argv, "cwprhus:e:q",
 				long_options, NULL)) != EOF) {
 		switch (c) {
 		case 'c':
@@ -279,6 +286,9 @@ int main(int argc, char **argv)
 		case 'q':
 			silent = true;
 			break;
+		case 'u':
+			opt_u = 1;
+			break;
 		default:
 			usage(argv[0]);
 			exit(1);
@@ -295,6 +305,37 @@ int main(int argc, char **argv)
 			rundir = "/run";
 		snprintf(psplash_pipe_path, sizeof(psplash_pipe_path), "%s/psplash_fifo", rundir);
 	}
+	if(opt_u) {
+		// init for HTTP PUT progress updates
+		curl_global_init(CURL_GLOBAL_ALL);
+ 
+		/* get a curl handle */
+		curl = curl_easy_init();
+		if(curl) {
+			// prepare header for PUT
+			put_headers = curl_slist_append(put_headers, "Content-Type: application/text");
+			if(put_headers == NULL) {
+				exit(1);
+			}
+			struct curl_slist *temp = curl_slist_append(put_headers, "charsets: utf-8");
+			if(temp == NULL) {
+				exit(1);
+			}
+			put_headers = temp;
+    		res = curl_easy_setopt(curl, CURLOPT_HTTPHEADER, put_headers);
+			if(res != CURLE_OK)
+				fprintf(stderr, "curl_easy_setopt() CURLOPT_HTTPHEADER failed: %s\n",
+						curl_easy_strerror(res));
+			res = curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "PUT");
+			if(res != CURLE_OK)
+				fprintf(stderr, "curl_easy_setopt() CURLOPT_CUSTOMREQUEST failed: %s\n",
+						curl_easy_strerror(res));
+			res = curl_easy_setopt(curl, CURLOPT_URL, "http://172.16.1.1:8280/progress");
+			if(res != CURLE_OK)
+				fprintf(stderr, "curl_easy_setopt() CURLOPT_URL failed: %s\n",
+						curl_easy_strerror(res));
+		}
+	}
 	connfd = -1;
 	redirected = !isatty(fileno(stdout));
 
@@ -410,6 +451,21 @@ int main(int argc, char **argv)
 				if (psplash_ok && ((msg.cur_step != curstep) || (msg.cur_percent != percent))) {
 					psplash_progress(psplash_pipe_path, &msg);
 				}
+				if (opt_u) {
+					int len = snprintf(put_buffer, 1024,
+						"#12,number: %d,step: %d,name: %s,percent: %d",
+						msg.nsteps,
+						msg.cur_step,
+						msg.cur_step ? msg.cur_image : "",
+						msg.cur_percent);
+					curl_easy_setopt(curl, CURLOPT_POSTFIELDS, put_buffer);
+					res = curl_easy_perform(curl);
+					/* Check for errors */
+					if(res != CURLE_OK)
+					fprintf(stderr, "curl_easy_perform() failed: %s\n",
+							curl_easy_strerror(res));
+				}
+
 				curstep = msg.cur_step;
 				percent = msg.cur_percent;
 			}
@@ -483,4 +539,10 @@ int main(int argc, char **argv)
 			break;
 		}
 	}
+
+	if(opt_u) {
+		/* always cleanup */
+		curl_easy_cleanup(curl);
+		curl_global_cleanup();
+	}
 }
